### PR TITLE
Update max allowed max min ratio

### PR DESCRIPTION
**Pull request by shibshib**
Param: maxallowedmaxminratio
Initial Value: 10
Suggested Value: 64
Time of Effect: Immediate

Background/Reasoning:
The original functionality of `normalize_max_multiple` and `maxallowedmaxminratio` was to reduce the power of outliers in the validation. The process of validation has now changed due to the synapse update. We are updating the parameter to reflect these changes

### CODE REVIEW

**Comment by mrseeker** [comment]
What will it do, who will get affected and why is it important to go that high instantly?

**Comment by opentaco** [comment]
### Recommendation: Remove `min_is_allowed_multiple_of_max` check in [weights.rs](https://github.com/opentensor/subtensor/blob/master/pallets/subtensor/src/weights.rs)

1. **Objective**: The real immediate objective should be to store validator's weight allocation `validator_weights` to the `chain_weights` as accurately as possible for the majority of nodes: `minimize[(chain_weights - validator_weights).abs()]`.
1. **Renormalization problem**: The first major issue is that validators currently only set weights for `min_allowed_weights` number of nodes, and that the chain normalizes these e.g. 600 weights to sum to 1. A validator needs multiple such renormalized weight-settings to start to cover all nodes, potentially overwriting previously set weights if the subtensor `step()` has not yet executed. This distorts the original `validator_weights` due to renormalization over subsets and due to overwriting.
1. **Solution**: Validators should simultaneously set their original `validator_weights` on chain in a single submission. This solves the renormalization problem that introduces very problematic distortion.
1. **min_max_ratio problem**: Exacerbating the renormalization problem is the additional application of `min_max_ratio` which `min_is_allowed_multiple_of_max` checks in [weights.rs](https://github.com/opentensor/subtensor/blob/master/pallets/subtensor/src/weights.rs). Each validator weight-setting fixes the `validator_weights.max()/validator_weights.min()=> min_max_ratio` which typically introduces a magnified distortion due to sensitivity to min/max outliers and due to significant alteration of the original `validator_weights`.

### Counter-arguments (reasons for applying `min_max_ratio`)

#### Adversarial resilience
1. Applying `min_max_ratio` could force validators to restrict the worst and best weight to limit a cabal from setting extreme weights that may favor them.
1. For example, a cabal cannot set all weight to themselves, because they have to set 600 weights and have to apply e.g. a 10:1 max:min ratio. So the chain prevents setting weights violating this condition.
1. However, the chain then makes a very strong assumption about the actual value contribution of participants in the network, and prevents validators from accurately recording their Shapley values on-chain.
1. By excluding part of the network not fitting the `min_max_ratio`, there could be no distortion. However, this will then exclude part of the network from being incentivized, when instead validators should fairly assess all nodes on-chain by setting weights for all nodes even near-zero weights.

#### Improved consensus
1. By restricting weight ratios and filtering weights the consensus algorithm could be more accurate in meeting the cabal decay property of `0.5 < [ SUM(i) incentive_i * ( 0.5 < consensus_i ) ]`, since weight magnitudes affect consensus_i.
1. Currently this affectation is mostly whether weights are non-zero, possibly whether weights exceed a threshold. Thus, by manipulating the min weight then the threshold and consensus can be manipulated in turn. However, this is a tenuous lever at best since the price of significant validator weight-setting error this introduces may not offset the adversarial resilience benefit. Keeping in mind that the real objective is adversarial resilience, cabal decay while fairly setting accurate weights that align with majority consensus.
1. Instead, a fully featured consensus algorithm allows for weights to be set freely and encourages weights set on the full network according to a fair global comparative evaluation. Then it relies simply on existing ownership via stake to preserve their share of new inflation by setting weights that are in consensus with majority stake.
1. Bittensor solves the nothing-at-stake problem by putting the share of new inflation at stake, such that out-of-consensus validators have new inflation slashed which decays their overall share. This ensures cabal decay and aligns validation to the majority ownership objectives.
1. A fully featured consensus algorithm does not need to influence weights, other than normalizing them and incentivizing their alignment with majority consensus.

### Simulating weight-setting and the min_max_ratio effect

#### Applying min_max_ratio and renormalization

1. We take typical network parameters of 150 validators, a representative set of actual validator Shapley values, and `min_allowed_weights = 600` to demonstrate the renormalization problem.
1. All validators randomly sample 600 Shapley values, renormalizes them, submits to a single representative on-chain weights set (note that reduction from 150 to just 10 weight-setting submissions preserves the conclusion), and then we compare with and without application of `min_max_ratio`.
1. We measure the absolute error of each individual node weight, with and without `min_max_ratio`.
1. Then we check how much we can reduce the error by removing `min_max_ratio` altogether. We plot the histogram of error reduction by removing `min_max_ratio`.

```python
validators = 150
min_allowed_weights = 600
# validator_weights = normalized Shapley values for all nodes

def weight_setting_error(min_max_ratio):
    chain_weights = torch.zeros_like(validator_weights)
    chain_weights_mmr = torch.zeros_like(validator_weights)  # min-max-ratio applied

    for v in range(validators):
        # validator sets min_allowed_weights
        sample = torch.randperm(len(validator_weights))[:min_allowed_weights]
        # randomly samples responsives to ensure fair weight setting
        weight_sample = validator_weights[sample].clone()

        chain_weights[sample] = weight_sample / weight_sample.sum()  # renormalize sample
        chain_weights_mmr[sample] = normalize_max_multiple(weight_sample, min_max_ratio)  # apply mmr
    
    # objective: minimize (chain_weights - validator_weights).abs()
    absolute_error = (chain_weights - validator_weights).abs()
    absolute_error_mmr = (chain_weights_mmr - validator_weights).abs()
    
    return absolute_error, absolute_error_mmr
```

1. We observe that the majority of nodes have a reduced weight-setting error when `min_max_ratio` is not applied. This corresponds to the majority mass on the negative side below 0. This is expected because of the amplified distortion of the original `validator_weights`.
1. However, for a minority of nodes there is a larger than usual error by not applying `min_max_ratio`.
1. The question is whether there are any advantages, like improved consensus or adversarial resilience, that outweigh the disadvantage of increasing weight-setting error for the majority of nodes.
1. Note also in the legend for different `min_max_ratio` that there is no setting large enough that alleviates the problem, so simply increasing `min_max_ratio` will generally not help due to the sensitivity problem.

Screenshot 2022-08-22 at 12 09 31

---

#### Validator check before applying `min_max_ratio`

1. Currently, the validator runs `normalize_max_multiple()` before every weight-setting, but it does not include a similar subtensor/weights.rs check to see if the condition is met. It always applies `min_max_ratio` in any case.
1. **Sensitivity problem**: Due to its sensitivity to min/max outliers, the exact ratio varies greatly, and there is simply no single `min_max_ratio` setting that will perfectly match the exact ratio.
1. **Solution**: By adding a check before, like below, we expect a reduction in weight-setting error because distortion is avoided.

```python
def normalize_max_multiple(  x: torch.FloatTensor, multiple:int = 3 ) -> 'torch.FloatTensor':
    r""" Normalizes the tensor x so that sum(x) = 1 and the max value is at most multiple times larger than the min value.
        Args:
            x (:obj:`torch.FloatTensor`):
                Tensor to be max_value normalized.
            multiple: float:
                Max value is multiple times larger than the min after normalization.     
        Returns:
            x (:obj:`torch.FloatTensor`):
                Normalized x tensor.
    """
    # ADD CHECK BEFORE (like subtensor/weights.rs)
    if x.max() / x.min() < multiple:
        return x / x.sum()

    x = x 
    shift = 1 / ( multiple - 1 )
    x = x - x.min()

    if x.sum() == 0:
        return torch.ones_like(x)/x.size(0)
    else:
        x = x / x.sum()
        y = (torch.tanh(x * len(x)) + shift)/(torch.tanh( x * len(x) ) + shift).sum()
        return y

```

We observe a larger error delta mass at 0, which means that for a very high `min_max_ratio` the check before can avoid application of distortion to the original values.

Screenshot 2022-08-22 at 12 09 51

---

#### Full weight-setting (no renormalization) with min_max_ratio

1. We can solve the renormalization problem by always setting weights for all nodes in a single weight-setting. 
1. Then, if we still apply `min_max_ratio` we see no redemption and it worsens all errors.

Screenshot 2022-08-22 at 12 42 23

**Review by opentaco — APPROVED** [review]

**Comment by Eugene-hu** [comment]
As @opentaco have said, the original functionality of `normalize_max_multiple` and thus `maxallowedmaxminratio` was to reduce the power of outliers in the validation. Previously, validators can become fixated on a single peer as it trains, resulting in extremely high weights for a select few peers and virtually no weights for the rest of the network. After the synapse update, this should no longer happen to the validators and thus, we should have a conversation about the use cases for `normalize_max_multiple`. There should be some form of normalization, but I am open to updating the function to something more sophisticated 

Additionally, the `normalize_max_multiple` was originally designed for the top 600 weights. Now as we are increasing the amount of weights that we are setting in #4, we should increase the min_max_ratio to accommodate for the changes.

**Review by Eugene-hu — APPROVED** [review]

**Review by isabella618033 — APPROVED** [review]

**Review by unconst — APPROVED** [review]

**Comment by shibshib** [comment]
Parameter updated on block: 2,102,096

explorer: https://explorer.nakamoto.opentensor.ai/#/explorer/query/0xecba2fd6f0b6fe8509026b78d81e1f4240275e98b51ecfb3203b16290d3b6953
